File: gtci/01-two-pointers/cpp/9-comparing-strings-containing-backspaces.cpp
#include <cassert>
#include <iostream>
#include <string>

using namespace std;

class Solution {
   public:
    static bool compare(const string &str1, const string &str2) {
        int idx1 = str1.size() - 1;
        int idx2 = str2.size() - 1;

        while (idx1 >= 0 || idx2 >= 0) {
            idx1 = Solution::getNextValidIndex(str1, idx1);
            idx2 = Solution::getNextValidIndex(str2, idx2);

            if (idx1 >= 0 && idx2 >= 0 && str1[idx1] != str2[idx2]) {
                return false;
            }

            idx1--;
            idx2--;
        }

        return idx1 == idx2;
    }

   private:
    static int getNextValidIndex(const string &str, int idx) {
        int backspaceCount = 0;

        while (idx >= 0) {
            if (str[idx] == '#') {
                backspaceCount++;
            } else if (backspaceCount > 0) {
                backspaceCount--;
            } else {
                return idx;
            }

            idx--;
        }

        return idx;
    }
};

int main() {
    Solution sol;

    assert(sol.compare("xy#z", "xzz#") == true);
    assert(sol.compare("xy#z", "xyz#") == false);
    assert(sol.compare("xp#", "xyz##") == true);
    cout << "All test cases passed." << endl;

    return 0;
}

File: gtci/01-two-pointers/python3/9-comparing-strings-containing-backspaces.py
class Solution:
    def compare(self, str1, str2):
        str1_ptr = len(str1) - 1
        str2_ptr = len(str2) - 1

        while str1_ptr >= 0 or str2_ptr >= 0:
            str1_ptr = self.get_next_valid_char(str1, str1_ptr)
            str2_ptr = self.get_next_valid_char(str2, str2_ptr)

            if str1_ptr >= 0 and str2_ptr >= 0 and str1[str1_ptr] != str2[str2_ptr]:
                return False

            str1_ptr -= 1
            str2_ptr -= 1

        return str1_ptr == str2_ptr

    def get_next_valid_char(self, s, i):
        backspace_count = 0
        while i >= 0:
            if s[i] == "#":
                backspace_count += 1
            elif backspace_count > 0:
                backspace_count -= 1
            else:
                return i

            i -= 1

        return i


if __name__ == "__main__":
    sol = Solution()
    assert sol.compare("xy#z", "xzz#") == True
    assert sol.compare("xy#z", "xyz#") == False
    assert sol.compare("xp#", "xyz##") == True
    print("All test cases passed.")
